File: test/lib/test_http.cpp
// Copyright (c) 2019,2021 <dev@brigid.jp>
// This software is released under the MIT License.
// https://opensource.org/licenses/mit-license.php

#include <brigid/http.hpp>
#include <brigid/version.hpp>
#include "test.hpp"

#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <memory>
#include <string>

namespace {
  using namespace std::placeholders;

  static const std::map<std::string, std::string> empty_header;

  class test_client {
  public:
    explicit test_client(
        bool credential = false,
        const std::string& username = std::string(),
        const std::string& password = std::string(),
        int cancel = 0)
      : session_(brigid::make_http_session(
            std::bind(&test_client::progress_cb, this, _1, _2),
            std::bind(&test_client::header_cb, this, _1, _2),
            std::bind(&test_client::write_cb, this, _1, _2),
            credential,
            username,
            password)),
        cancel_(cancel),
        cancel_count_(),
        progress_count_(),
        header_count_(),
        write_count_(),
        code_(),
        canceled_() {}

    int request(
        const std::string& method,
        const std::string& url,
        const std::map<std::string, std::string>& header = empty_header,
        brigid::http_request_body body = brigid::http_request_body::none,
        const char* data = nullptr,
        size_t size = 0) {
      cancel_count_ = 0;
      progress_count_ = 0;
      header_count_ = 0;
      write_count_ = 0;
      code_ = 0;
      canceled_ = !session_->request(method, url, header, body, data, size);
      body_ = out_.str();
      out_.str(std::string());
      out_.clear();
      return code_;
    }

    size_t cancel_count() const {
      return cancel_count_;
    }

    size_t progress_count() const {
      return progress_count_;
    }

    size_t header_count() const {
      return header_count_;
    }

    size_t write_count() const {
      return write_count_;
    }

    int code() const {
      return code_;
    }

    const std::map<std::string, std::string>& header() const {
      return header_;
    }

    std::string header(const std::string& key) const {
      auto iterator = header_.find(key);
      if (iterator == header_.end()) {
        return std::string();
      } else {
        return iterator->second;
      }
    }

    const std::string& body() const {
      return body_;
    }

    bool canceled() const {
      return canceled_;
    }

  private:
    std::unique_ptr<brigid::http_session> session_;
    int cancel_;
    size_t cancel_count_;
    size_t progress_count_;
    size_t header_count_;
    size_t write_count_;
    int code_;
    std::map<std::string, std::string> header_;
    std::ostringstream out_;
    std::string body_;
    bool canceled_;

    bool progress_cb(size_t now, size_t total) {
      std::cout << "[debug] progress " << now << "/" << total << "\n";

      if (cancel_ == 1) {
        if (now * 2 >= total) {
          ++cancel_count_;
          std::cout << "explicitly cancel in the progress callback\n";
          return false;
        }
      }

      ++progress_count_;
      std::cout << "progress " << now << "/" << total << "\n";
      return true;
    }

    bool header_cb(int code, const std::map<std::string, std::string>& header) {
      if (cancel_ == 2) {
        ++cancel_count_;
        std::cout << "explicitly cancel in the header callback\n";
        return false;
      }

      ++header_count_;
      code_ = code;
      header_ = header;
      std::cout << "header " << code_ << "\n";
      return true;
    }

    bool write_cb(const char* data, size_t size) {
      if (cancel_ == 3) {
        ++cancel_count_;
        std::cout << "explicitly cancel in the writer callback\n";
        return false;
      }

      ++write_count_;
      out_.write(data, size);
      return true;
    }
  };

  void test1() {
    test_client client;
    client.request("GET", "https://brigid.jp/");
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.header_count() == 1);
    BRIGID_CHECK(client.write_count() == 0);
    BRIGID_CHECK(client.header("Content-Length") == "0");
    BRIGID_CHECK(client.header("Content-Type") == "text/html; charset=UTF-8");
    BRIGID_CHECK(client.body().empty());
  }

  void test2() {
    static const std::string data = "foo\nbar\nbaz\nqux\n";

    test_client client;
    client.request("PUT", "https://brigid.jp/test/dav/auth-none/test.txt", empty_header, brigid::http_request_body::data, data.data(), data.size());
    BRIGID_CHECK(client.progress_count() > 0);
    BRIGID_CHECK(client.header_count() == 1);
    BRIGID_CHECK(client.code() == 201 || client.code() == 204);
    if (client.code() == 204) {
      BRIGID_CHECK(client.write_count() == 0);
      BRIGID_CHECK(client.body().empty());
    }

    client.request("GET", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.header_count() == 1);
    BRIGID_CHECK(client.write_count() > 0);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body() == data);

    client.request("HEAD", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.write_count() == 0);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().empty());

    client.request("DELETE", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.write_count() == 0);
    BRIGID_CHECK(client.code()== 204);
    BRIGID_CHECK(client.body().empty());

    client.request("GET", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.write_count() > 0);
    BRIGID_CHECK(client.code() == 404);

    client.request("HEAD", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.write_count() == 0);
    BRIGID_CHECK(client.code() == 404);
    BRIGID_CHECK(client.body().empty());
  }

  void test3() {
    std::string filename = "test.dat";
    std::string data;
    for (size_t i = 0; i < 4096; i += 16) {
      data += "0123456789ABCDE\n";
    }
    {
      std::ofstream out(filename.c_str(), std::ios::out | std::ios::binary);
      for (size_t i = 0; i < 1024 * 1024; i += data.size()) {
        out << data;
      }
    }

    test_client client;
    client.request("PUT", "https://brigid.jp/test/dav/auth-none/test.txt", empty_header, brigid::http_request_body::file, filename.data(), filename.size());
    BRIGID_CHECK(client.progress_count() > 0);
    BRIGID_CHECK(client.code() == 201 || client.code() == 204);

    std::cout << "progress_count " << client.progress_count() << " " << 1024 * 1024 / client.progress_count() << "\n";

    client.request("GET", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().size() == 1024 * 1024);

    std::cout << "write_count " << client.write_count() << " " << 1024 * 1024 / client.write_count() << "\n";

    client.request("DELETE", "https://brigid.jp/test/dav/auth-none/test.txt");
    BRIGID_CHECK(client.progress_count() == 0);
    BRIGID_CHECK(client.code()== 204);
    BRIGID_CHECK(client.body().empty());

    remove(filename.c_str());
  }

  void test4() {
    test_client client;
    client.request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=1");
    BRIGID_CHECK(client.header_count() == 1);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body() == "ok\n");
    BRIGID_CHECK(client.header("Location") == "");

    client.request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=11");
    BRIGID_CHECK(client.header_count() == 1);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body() == "ok\n");
    BRIGID_CHECK(client.header("Location") == "");

    BRIGID_CHECK_THROW([&](){ client.request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=22"); });
  }

  void test5() {
    {
      test_client client;
      client.request("GET", "https://brigid.jp/test/dav/auth-none/");
      BRIGID_CHECK(client.code() == 200);

      client.request("GET", "https://brigid.jp/test/dav/auth-basic/");
      BRIGID_CHECK(client.code() == 401);
    }

    {
      test_client client(true, "brigid", "O6jIOchrWCGuOSB4");
      client.request("GET", "https://brigid.jp/test/dav/auth-none/");
      BRIGID_CHECK(client.code() == 200);

      client.request("GET", "https://brigid.jp/test/dav/auth-basic/");
      BRIGID_CHECK(client.code() == 200);
    }
  }

  void test6() {
    {
      test_client client;
      client.request("GET", "https://brigid.jp/test/dav/auth-none/");
      BRIGID_CHECK(client.code() == 200);

      client.request("GET", "https://brigid.jp/test/dav/auth-digest/");
      BRIGID_CHECK(client.code() == 401);
    }

    {
      test_client client(true, "brigid", "YlrMTunTORZvrgSt");
      client.request("GET", "https://brigid.jp/test/dav/auth-none/");
      BRIGID_CHECK(client.code() == 200);

      client.request("GET", "https://brigid.jp/test/dav/auth-digest/");
      BRIGID_CHECK(client.code() == 200);
    }
  }

  void test7() {
    static const std::string data = R"({"foo":42})";

    test_client client;
    std::map<std::string, std::string> header {
      { "Content-Type", "application/json; charset=UTF-8" },
    };

    client.request("POST", "https://brigid.jp/test/cgi/cat.cgi", header, brigid::http_request_body::data, data.data(), data.size());
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body() == data);

    client.request("POST", "https://brigid.jp/test/cgi/env.cgi", header, brigid::http_request_body::data, data.data(), data.size());
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().find("CONTENT_TYPE=application/json; charset=UTF-8\n") != std::string::npos);
  }

  void test8() {
    test_client client;
    BRIGID_CHECK_THROW([&](){ client.request("GET", "https://133.242.153.239/"); });
  }

  void test9() {
    test_client client;
    client.request("GET", "https://brigid.jp/test/cgi/nph-header.cgi");
    BRIGID_CHECK(client.code() == 200);

    for (const auto& field : client.header()) {
      std::cout << "[" << field.first << "]=[" << field.second << "]\n";
    }

    auto check = [](const std::string& value) -> bool {
      size_t n = value.size();
      return n > 6
          && value.substr(0, 3) == "foo"
          && value.substr(3, n - 6).find_first_not_of(" \t") == std::string::npos
          && value.substr(n - 3, 3) == "bar";
    };

    BRIGID_CHECK(client.header().size() == 11);
    BRIGID_CHECK(client.header("X-Test1") == "foo bar");
    BRIGID_CHECK(client.header("X-Test2") == "foo bar");
    BRIGID_CHECK(client.header("X-Test3") == "foo  bar");
    BRIGID_CHECK(check(client.header("X-Test4")));
    BRIGID_CHECK(check(client.header("X-Test5")));
    BRIGID_CHECK(client.header("X-Test6") == "foo bar");
    BRIGID_CHECK(client.header("x-test7") == "foo bar");
    BRIGID_CHECK(client.header("X-tEsT8") == "foo bar");
  }

  void test10() {
    static const std::string data = "";

    test_client client;
    std::map<std::string, std::string> header {
      { "Content-Type", "application/x-www-form-urlencoded" },
    };

    client.request("POST", "https://brigid.jp/test/cgi/cat.cgi", header, brigid::http_request_body::data, data.data(), data.size());
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().empty());

    client.request("POST", "https://brigid.jp/test/cgi/env.cgi", header, brigid::http_request_body::data, data.data(), data.size());
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().find("CONTENT_TYPE=application/x-www-form-urlencoded\n") != std::string::npos);
  }

  void test11() {
    test_client client;
    BRIGID_CHECK_THROW([&](){ client.request("GET", "https://no-such-host.brigid.jp/"); });
  }

  void test12() {
    test_client client;
    BRIGID_CHECK_THROW([&](){ client.request("GET", "!!! invalid url !!!"); });
  }

  void test13() {
    test_client client;
    client.request("GET", "https://brigid.jp/test/%63%67%69/env.cgi?%20%21");
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().find("QUERY_STRING=%20%21\n") != std::string::npos);
  }

  void test14() {
    test_client client;
    client.request("GET", "https://brigid.jp/test/cgi/cookie.cgi");
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.header("Set-Cookie") == "code=42; Path=/test/cgi/; Secure");
    BRIGID_CHECK(client.body() == "HTTP_COOKIE=\n");

    client.request("GET", "https://brigid.jp/test/cgi/cookie.cgi");
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.header("Set-Cookie") == "code=42; Path=/test/cgi/; Secure");
    BRIGID_CHECK(client.body() == "HTTP_COOKIE=\n");

    std::map<std::string, std::string> header {
      { "Cookie", "code=69" },
    };
    client.request("GET", "https://brigid.jp/test/cgi/cookie.cgi", header);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.header("Set-Cookie") == "");
    BRIGID_CHECK(client.body() == "HTTP_COOKIE=code=69\n");
  }

  void test15() {
    test_client client;

    std::string ua = std::string("brigid/") + brigid::get_version();
    std::map<std::string, std::string> header {
      { "User-Agent", ua },
    };

    client.request("GET", "https://brigid.jp/test/cgi/env.cgi", header);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.body().find("USER_AGENT=" + ua + "\n") != std::string::npos);
  }

  void test16() {
    std::string filename = "test.dat";
    std::string data;
    for (size_t i = 0; i < 4096; i += 16) {
      data += "0123456789ABCDE\n";
    }
    {
      std::ofstream out(filename.c_str(), std::ios::out | std::ios::binary);
      for (size_t i = 0; i < 8 * 1024 * 1024; i += data.size()) {
        out << data;
      }
    }

    test_client client(false, "", "", 1);;
    client.request("PUT", "https://brigid.jp/test/dav/auth-none/test.txt", empty_header, brigid::http_request_body::file, filename.data(), filename.size());
    std::cout << "[debug] progress_count " << client.progress_count() << "\n";
    BRIGID_CHECK(client.progress_count() > 0);
    std::cout << "[debug] cancel_count " << client.cancel_count() << "\n";
    BRIGID_CHECK(client.cancel_count() == 1);
    BRIGID_CHECK(client.code() == 0);
    BRIGID_CHECK(client.canceled());

    remove(filename.c_str());
  }

  void test17() {
    test_client client(false, "", "", 2);
    client.request("GET", "https://brigid.jp/test/cgi/env.cgi");
    BRIGID_CHECK(client.cancel_count() == 1);
    BRIGID_CHECK(client.code() == 0);
    BRIGID_CHECK(client.canceled());
  }

  void test18() {
    test_client client(false, "", "", 3);
    client.request("GET", "https://brigid.jp/test/cgi/env.cgi");
    BRIGID_CHECK(client.cancel_count() == 1);
    BRIGID_CHECK(client.code() == 200);
    BRIGID_CHECK(client.canceled());
  }

  BRIGID_MAKE_TEST_CASE(test1);
  BRIGID_MAKE_TEST_CASE(test2);
  BRIGID_MAKE_TEST_CASE(test3);
  BRIGID_MAKE_TEST_CASE(test4);
  BRIGID_MAKE_TEST_CASE(test5);
  BRIGID_MAKE_TEST_CASE(test6);
  BRIGID_MAKE_TEST_CASE(test7);
  BRIGID_MAKE_TEST_CASE(test8);
  BRIGID_MAKE_TEST_CASE(test9);
  BRIGID_MAKE_TEST_CASE(test10);
  BRIGID_MAKE_TEST_CASE(test11);
  BRIGID_MAKE_TEST_CASE(test12);
  BRIGID_MAKE_TEST_CASE(test13);
  BRIGID_MAKE_TEST_CASE(test14);
  BRIGID_MAKE_TEST_CASE(test15);
  BRIGID_MAKE_TEST_CASE(test16);
  BRIGID_MAKE_TEST_CASE(test17);
  BRIGID_MAKE_TEST_CASE(test18);
}

File: test/lua/test_http.lua
-- Copyright (c) 2019,2021 <dev@brigid.jp>
-- This software is released under the MIT License.
-- https://opensource.org/licenses/mit-license.php

local brigid = require "brigid"
local test_http_client = require "test_http_client"
local test_suite = require "test_suite"

local data4k = ("0123456789ABCDE\n"):rep(4096 / 16)

local suite = test_suite "test_http"

local client

function suite:test_create_data()
  local out = assert(io.open(test_cwd .. "/test.dat", "wb"))
  for i = 1, 1024 * 1024 / #data4k do
    out:write(data4k)
  end
  out:close()
end

function suite:test_open_client_noauth()
  client = test_http_client()
end

function suite:test1()
  local code, header, body = client:request("GET", "https://brigid.jp/")
  assert(code == 200)
  assert(header["Content-Length"] == "0")
  assert(header["Content-Type"] == "text/html; charset=UTF-8")
  assert(body == "")
end

local data = "foo\nbar\nbaz\nqux\n";

function suite:test2()
  local code, header, body = client:request_data("PUT", "https://brigid.jp/test/dav/auth-none/test.txt", nil, data)
  assert(code == 201 or code == 204)
  if code == 204 then
    assert(body == "")
  end
end

function suite:test3()
  local code, header, body = client:request("GET", "https://brigid.jp/test/dav/auth-none/test.txt");
  assert(code == 200)
  assert(body == data)
end

function suite:test4()
  local code, header, body = client:request("HEAD", "https://brigid.jp/test/dav/auth-none/test.txt");
  assert(code == 200)
  assert(body == "")
end

function suite:test5()
  local code, header, body = client:request("DELETE", "https://brigid.jp/test/dav/auth-none/test.txt");
  assert(code == 204)
  assert(body == "")
end

function suite:test6()
  local code, header, body = client:request("GET", "https://brigid.jp/test/dav/auth-none/test.txt")
  assert(code == 404)
end

function suite:test7()
  local code, header, body = client:request("HEAD", "https://brigid.jp/test/dav/auth-none/test.txt")
  assert(code == 404)
  assert(body == "")
end

function suite:test8()
  local code, header, body = client:request_file("PUT", "https://brigid.jp/test/dav/auth-none/test.txt", nil, test_cwd .. "/test.dat")
  assert(code == 201 or code == 204)
end

function suite:test9()
  local code, header, body = client:request_file("GET", "https://brigid.jp/test/dav/auth-none/test.txt")
  assert(code == 200)
  assert(body == data4k:rep(1024 * 1024 / #data4k))
end

function suite:test10()
  local code, header, body = client:request_file("DELETE", "https://brigid.jp/test/dav/auth-none/test.txt")
  assert(code == 204)
  assert(body == "")
end

function suite:test11()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=1")
  assert(code == 200)
  assert(body == "ok\n")
  assert(not header.Locatuion)
end

function suite:test12()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=11")
  assert(code == 200)
  assert(body == "ok\n")
  assert(not header.Locatuion)
end

function suite:test13()
  local result, message = client:request("GET", "https://brigid.jp/test/cgi/redirect.cgi?count=22")
  print(message)
  assert(not result)
end

function suite:test14()
  local code, header, body = client:request("GET", "https://brigid.jp/test/dav/auth-basic/")
  assert(code == 401)
end

function suite:test_close_client_noauth()
  client:close()
  local result, message = pcall(function ()
    client:request("GET", "https://brigid.jp/")
  end)
  print(message)
  assert(not result)
  client = nil
end

function suite:test_open_client_auth()
  client = test_http_client("brigid", "O6jIOchrWCGuOSB4")
end

function suite:test15()
  local code, header, body = client:request("GET", "https://brigid.jp/test/dav/auth-basic/")
  assert(code == 200)
end

function suite:test_close_client_auth()
  client:close()
  client = nil
end

function suite:test_open_client()
  client = test_http_client()
end

local data = [[{"foo":42}]]
local request_header = { ["Content-Type"] = "application/json; charset=UTF-8" }

function suite:test16()
  local code, header, body = client:request_data("POST", "https://brigid.jp/test/cgi/cat.cgi", request_header, data)
  assert(code == 200)
  assert(body == data)
end

function suite:test17()
  local code, header, body = client:request_data("POST", "https://brigid.jp/test/cgi/env.cgi", request_header, data)
  assert(code == 200)
  assert(body:find("CONTENT_TYPE=application/json; charset=UTF-8\n", 1, true))
end

function suite:test18()
  local result, message = client:request("GET", "https://133.242.153.239/")
  print(message)
  assert(not result)
end

function suite:test19()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/nph-header.cgi")
  assert(code == 200)

  for k, v in pairs(header) do
    print("[" .. k .. "]=[" .. v .. "]")
  end

  assert(header["X-Test1"] == "foo bar")
  assert(header["X-Test2"] == "foo bar")
  assert(header["X-Test3"] == "foo  bar")
  assert(header["X-Test4"]:find "^foo[ \t]+bar$")
  assert(header["X-Test5"]:find "^foo[ \t]+bar$")
  assert(header["X-Test6"] == "foo bar")
  assert(header["x-test7"] == "foo bar")
  assert(header["X-tEsT8"] == "foo bar")
end

local data = ""
local request_header = { ["Content-Type"] = "application/x-www-form-urlencoded" }

function suite:test20()
  local code, header, body = client:request_data("POST", "https://brigid.jp/test/cgi/cat.cgi", request_header, data)
  assert(code == 200)
  assert(body == "")
end

function suite:test21()
  local code, header, body = client:request_data("POST", "https://brigid.jp/test/cgi/env.cgi", request_header, data)
  assert(code == 200)
  assert(body:find("CONTENT_TYPE=application/x-www-form-urlencoded\n", 1, true))
end

function suite:test22()
  local result, message = client:request("GET", "https://no-such-host.brigid.jp/")
  print(message)
  assert(not result)
end

function suite:test23()
  local result, message = client:request("GET", "!!! invalid url !!!")
  print(message)
  assert(not result)
end

function suite:test24()
  local code, header, body = client:request("POST", "https://brigid.jp/test/%63%67%69/env.cgi?%20%21")
  assert(code == 200)
  assert(body:find("QUERY_STRING=%20%21\n", 1, true))
end

function suite:test25()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/cookie.cgi")
  assert(code == 200)
  assert(header["Set-Cookie"] == "code=42; Path=/test/cgi/; Secure")
  assert(body == "HTTP_COOKIE=\n")
end

function suite:test26()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/cookie.cgi")
  assert(code == 200)
  assert(header["Set-Cookie"] == "code=42; Path=/test/cgi/; Secure")
  assert(body == "HTTP_COOKIE=\n")
end

function suite:test27()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/cookie.cgi", { Cookie = "code=69" })
  assert(code == 200)
  assert(not header["Set-Cookie"])
  assert(body == "HTTP_COOKIE=code=69\n")
end

function suite:test28()
  local ua = "brigid/" .. brigid.get_version()
  local code, header, body = client:request("GET", "https://brigid.jp/test/cgi/env.cgi", { ["User-Agent"] = ua })
  assert(code == 200)
  assert(body:find("USER_AGENT=" .. ua .. "\n", 1, true))
end

function suite:test_close_client()
  assert(client.session:close())
  assert(client.session:close())
  client = nil
end

function suite:test29()
  -- explicitly cancel in the progress callback
  -- NSURLSession backend is not canceled immediately

  local canceling = 0

  local session = brigid.http_session {
    progress = function (now, total)
      print(("progress %5.1f%%"):format(now * 100 / total))
      if now * 2 >= total then
        print "explicitly cancel in the progress callback"
        canceling = canceling + 1
        return false
      end
    end;

    header = function ()
      canceling = canceling + 1
    end;

    write = function ()
      canceling = canceling + 1
    end;
  }
  local result, message = session:request {
    method = "PUT",
    url = "https://brigid.jp/test/dav/auth-none/test.txt";
    file = test_cwd .. "/test.dat";
  }
  print(message, canceling)
  assert(not result)
  assert(message == "canceled")
  assert(canceling == 1)
end

function suite:test_remove_data()
  os.remove(test_cwd .. "/test.dat")
end

return suite
